### PR TITLE
Ingress controller tests

### PR DESCRIPTION
**Pull request by bprashanth**
A basic suite of tests existing/new controllers should pass to claim to have satisfied an Ingress.

### CODE REVIEW

**Comment by bprashanth** [comment]
@quinton-hoole assigned to you since you have context from previous review, feel free to re-assign. I renamed serviceloadbalancer -> ingress_conformance, so you can look at the second commit for an actual diff.

**Comment by quinton-hoole** [comment]
Sure, I can take a look @bprashanth

File: test/e2e/testing-manifests/nginxtest/rc.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: nginx-test
+    spec:
+      containers:
+      - name: nginx-test
+        image: bprashanth/nginx-test:1.0
```
Comment (bprashanth) [inline test/e2e/testing-manifests/nginxtest/rc.yaml]: need to upload to gcr.io, this is the nginx pod included in this pr

File: test/e2e/testing-manifests/glbc/rc.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  template:
+    metadata:
+      labels:
+        app: glbc
+        version: v1
+    spec:
+      # It only needs this long if started with --delete-all-on-quit.
+      terminationGracePeriodSeconds: 600
+      containers:
+      - image: bprashanth/glbc:0.0
```
Comment (bprashanth) [inline test/e2e/testing-manifests/glbc/rc.yaml]: need to upload to gcr.io, this is the gce controller being tested https://github.com/kubernetes/contrib/tree/master/Ingress/controllers/gce

File: test/e2e/testing-manifests/glbc/default-svc-rc.yaml [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
     metadata:
       labels:
-        app: nginx
+        app: default-backend
     spec:
       containers:
-      - name: nginxhttps
-        image: bprashanth/nginxhttps:1.0
+      - name: default-backend
+        image: bprashanth/echoserver:0.0
```
Comment (bprashanth) [inline test/e2e/testing-manifests/glbc/default-svc-rc.yaml]: will replace this container with something on gcr.io, it's just a plance holder for a 404 page

**Comment by bprashanth** [comment]
The jenkins failure is weird, i would think jenkins would either not touch the new code at all, or grab all of it, but it seems to have run the go part and left out the .yaml files:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/ingress_conformance.go:273
Expected error:
    <*os.PathError | 0xc2083d1200>: {
        Op: "open",
        Path: "/jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@2/kubernetes/test/e2e/testing-manifests/glbc/default-svc-rc.yaml",
        Err: 0x2,
    }
    open /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@2/kubernetes/test/e2e/testing-manifests/glbc/default-svc-rc.yaml: no such file or directory
not to have occurred
```

**Comment by bprashanth** [comment]
Previous test was a little too ambitious, I've watered it down to just testing the GCE controller as an addon (https://github.com/kubernetes/kubernetes/pull/16232)

**Comment by ashcrow** [comment]
Timeout happened at https://github.com/bprashanth/kubernetes/blob/IngressE2E/test/e2e/ingress.go#L180

```
	• Failure [307.921 seconds]
	glbc
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/ingress.go:221
	  glbc [It]
	  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/ingress.go:220
	
	  Expected error:
	      <*errors.errorString | 0xc208164340>: {
	          s: "timed out waiting for the condition",
	      }
	      timed out waiting for the condition
	  not to have occurred
	
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/ingress.go:180
```

**Comment by bprashanth** [comment]
Thanks, that's expected, this test depends on https://github.com/kubernetes/kubernetes/pull/16232 which actually starts the loadbalancer controller as a cluster addon to satisfy these requests for ingress

**Comment by ashcrow** [comment]
@bprashanth ah ok. So this shouldn't be merged until #16232 is merged?

File: test/e2e/ingress.go [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+//   - L7 is sending traffic to the default backend of the addon.
+//   - Backend is receiving /foo when it expects /bar.
+// * Repeated 5xx:
+//   - Out of quota (describe ing should show you if this is the case)
+//   - Mismatched service/container port, or endpoints are dead.
+
+var (
+	appPrefix         = "foo-app-"
+	pathPrefix        = "foo"
+	testImage         = "bprashanth/n-way-http:1.0"
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Can we put this image in a non-personal place?

File: test/e2e/ingress.go [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+	// Average duration of loadbalancer startup.
+	avgLBLatency = 7 * time.Minute
+
+	// On average it takes ~6 minutes for a single backend to come online.
+	// We *don't* expect this poll to consistently take 15 minutes for every
+	// Ingress as GCE is creating/checking backends in parallel, but at the
+	// same time, we're not testing GCE startup latency. So give it enough
+	// time, and fail if the average is too high.
+	lbPollTimeout  = 15 * time.Minute
+	lbPollInterval = 30 * time.Second
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: You might consider reducing this.  5sec?  That would shave 15 seconds off the running time of this test, on average.
With 200 tests, every 15 sec counts :-)

File: test/e2e/ingress.go [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+	}
+	Logf("Creating ingress %v", start)
+	_, err := c.Extensions().Ingress(ns).Create(&ing)
+	Expect(err).NotTo(HaveOccurred())
+	return ing
+}
+
+// createApp will create a single RC and Svc. The Svc will match pods of the
+// RC using the selector: 'name'=<name arg>
+func createApp(c *client.Client, ns string, i int) {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: nit: This seems like a very useful utility for more than just this test.  Suggest factoring it out into e2e/util library.  In the process you'll want to improve its decoupling by passing in appPrefix, httpContainerPort, testImage etc as parameters.
Happy to do all this in a follow-on PR if you prefer.  Just create an issue to track that.

File: test/e2e/ingress.go [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+		fmt.Sprintf("--start=%d", i),
+		fmt.Sprintf("--prefix=%v", pathPrefix),
+		fmt.Sprintf("--port=%d", httpContainerPort),
+	}
+	_, err = c.ReplicationControllers(ns).Create(rc)
+	Expect(err).NotTo(HaveOccurred())
+}
+
+// gcloudUnmarshal unmarshals json output of gcloud into given out interface.
+func gcloudUnmarshal(resource, regex string, out interface{}) {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Similar to above comment, this would make a great library function in e2e/util

File: test/e2e/ingress.go [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+	if len(fwList) != 0 {
+		for _, f := range fwList {
+			msg += fmt.Sprintf("%v\n", f.Name)
+		}
+		return fmt.Errorf("Found forwarding rules:\n%v", msg)
+	}
+	return nil
+}
+
+var _ = Describe("GLBC", func() {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Full name please. Oncalls are unlikely to know what a GLBC is.  Here's one suggestion:
Describe("GCE L7 Load Balancer"....
It("should respond to GET across all ingresses and all rules")

... or something like that.

File: test/e2e/ingress.go [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+	AfterEach(func() {
+		framework.afterEach()
+		wait.Poll(lbPollInterval, lbPollTimeout, func() (bool, error) {
+			if err := checkLeakedResources(); err != nil {
+				Logf("Still waiting for glbc to cleanup: %v", err)
+				return false, nil
+			}
+			return true, nil
+		})
+
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Perhaps log success here?

File: test/e2e/ingress.go [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+		// In most cases slow loadbalancer creation/startup translates directly to
+		// GCE api sluggishness. However this might be because of something the
+		// controller is doing, eg: maxing out QPS by repeated polling.
+		sort.Sort(timeSlice(creationTimes))
+		perc50 := creationTimes[len(creationTimes)/2]
+		if perc50 > avgLBLatency {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: nit: You're mixing averages with 50th percentiles here.  This could lead to confusion.  Also, "LBLatency" implies the latency introduced by the loadbalancer itself when directing traffic. Rename avgLBLatency to expectedLBCreationTime ?

File: test/e2e/ingress.go [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+			Failf("Average creation time is too high: %+v", creationTimes)
+		}
+		By(fmt.Sprintf("Loadbalancer creation latencies %+v", creationTimes))
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: I'd suggest standardising on the use of either "creation latency" or "creation time" across all of this code, but not mixing both terms.

File: test/e2e/ingress.go [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+
+		sort.Sort(timeSlice(responseTimes))
+		perc50 = responseTimes[len(responseTimes)/2]
+		if perc50 > avgLBLatency {
+			Failf("Average startup time is too high: %+v", responseTimes)
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: As per above comment.  You've introduced a third term ("startup time") for the same thing?

Comment (quinton-hoole) [inline test/e2e/ingress.go]: Sorry, this is clearly something different, so ignore the above comment.  But it's confusing that you re-use avgLBLatency here again.  Perhaps introduce a second limit, or name it more clearly?

Comment (bprashanth) [inline test/e2e/ingress.go]: im just clubbing everything related to how long gce takes to do X for the loadbalacer setup into one constant (X could be get ip, create healthcheck, start checking backends etc). I can rename to something like GCEWatiTimeout, as this *should* be 100% gce but there's a chance we're contributing to it by repeatedly polling or something in a way that maxes out qps.

File: test/e2e/ingress.go [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+				return false, nil
+			}
+			return true, nil
+		})
+
+		// In most cases slow loadbalancer creation/startup translates directly to
+		// GCE api sluggishness. However this might be because of something the
+		// controller is doing, eg: maxing out QPS by repeated polling.
+		sort.Sort(timeSlice(creationTimes))
+		perc50 := creationTimes[len(creationTimes)/2]
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: It feels kinda weird to insert what is essentially a performance test into AfterEach(), which is intended primarily for teardown.  But I understand that it's difficult to put it in a real "It()" type test without either polluting the existing functionality test, or wasting time setting up a second load balancer just to do the performance test.  Perhaps a reasonable middle-ground is to factor this performance test code into en explicit test function, and perhaps use Expect() rather than your if statement?

File: test/e2e/ingress.go [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+		// Create numApps apps, exposed via numIng Ingress each with 2 paths.
+		// Eg with numApp=10, numIng=5:
+		// apps: {foo-app-(0-10)}
+		// ingress: {foo-app-(0, 2, 4, 6, 8)}
+		// paths:
+		//  ingress foo-app-0:
+		//	  default1.bar.com
+		//	  foo0.bar.com: /foo0
+		//	  foo1.bar.com: /foo1
+		if numApps < numIng {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Gomega assertions are slightly more terse, and arguably more standard:
https://onsi.github.io/gomega/#making-assertions
https://onsi.github.io/gomega/#provided-matchers
But I don't feel too strongly about this.  Just an idea.

File: test/e2e/ingress.go [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+//   - Mismatched service/container port, or endpoints are dead.
+
+var (
+	appPrefix         = "foo-app-"
+	pathPrefix        = "foo"
+	testImage         = "bprashanth/n-way-http:1.0"
+	httpContainerPort = 8080
+
+	// Average duration of loadbalancer startup.
+	avgLBLatency = 7 * time.Minute
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: See comments below.  I think that this needs to be renamed.

File: test/e2e/ingress.go [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+	// We *don't* expect this poll to consistently take 15 minutes for every
+	// Ingress as GCE is creating/checking backends in parallel, but at the
+	// same time, we're not testing GCE startup latency. So give it enough
+	// time, and fail if the average is too high.
+	lbPollTimeout  = 15 * time.Minute
+	lbPollInterval = 30 * time.Second
+
+	// One can scale this test by tweaking numApps and numIng, the former will
+	// create more RCs/Services and add them to a single Ingress, while the latter
+	// will create smaller, more fragments Ingresses.
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: typo on fragments?

File: test/e2e/ingress.go [hunk 17]
```diff
@@ -0,0 +1,311 @@
...
+	numApps = 6
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: It might be worth mentioning in a comment here why you chose these numbers, in case someone comes along in the future saying "this is too slow - let me reduce these" without understanding the implications properly.

File: test/e2e/ingress.go [hunk 18]
```diff
@@ -0,0 +1,311 @@
...
+	_, err = c.ReplicationControllers(ns).Create(rc)
+	Expect(err).NotTo(HaveOccurred())
+}
+
+// gcloudUnmarshal unmarshals json output of gcloud into given out interface.
+func gcloudUnmarshal(resource, regex string, out interface{}) {
+	output, err := exec.Command("gcloud", "compute", resource, "list",
+		fmt.Sprintf("--regex=%v", regex), "-q", "--format=json").CombinedOutput()
+	if err != nil {
+		Failf("%v", err)
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Prefix the error message with some context?  "Error executing gcloud list" or something like that?

File: test/e2e/ingress.go [hunk 19]
```diff
@@ -0,0 +1,311 @@
...
+	}
+	if err := json.Unmarshal([]byte(output), out); err != nil {
+		Failf("%v", err)
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Similar to above comment.  Prefix with "Error unmarshalling gcloud command output"?

File: test/e2e/ingress.go [hunk 20]
```diff
@@ -0,0 +1,311 @@
...
+	}
+}
+
+func checkLeakedResources() error {
+	msg := ""
+	// TODO: These are the main resources but we should really check them all.
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Please create a github issue, and reference it here.

File: test/e2e/ingress.go [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+	beList := []compute.BackendService{}
+	gcloudUnmarshal("backend-services", "k8s-be-[0-9]+", &beList)
+	if len(beList) != 0 {
+		for _, b := range beList {
+			msg += fmt.Sprintf("%v\n", b.Name)
+		}
+		return fmt.Errorf("Found backend services:\n%v", msg)
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Please make the error message more obvious, e.g. "Failed to clean up cloud resources - expected none but found backend services ...."
Gomega assertions might be convenient here:
https://onsi.github.io/gomega/#making-assertions
https://onsi.github.io/gomega/#provided-matchers

File: test/e2e/ingress.go [hunk 22]
```diff
@@ -0,0 +1,311 @@
...
+	}
+	fwList := []compute.ForwardingRule{}
+	gcloudUnmarshal("forwarding-rules", "k8s-fw-.*", &fwList)
+	if len(fwList) != 0 {
+		for _, f := range fwList {
+			msg += fmt.Sprintf("%v\n", f.Name)
+		}
+		return fmt.Errorf("Found forwarding rules:\n%v", msg)
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Same comment as above.

File: test/e2e/ingress.go (hunk 13 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Similar to your checkLeakedResources() function.

File: test/e2e/ingress.go [hunk 23]
```diff
@@ -0,0 +1,311 @@
...
+		//  ingress foo-app-0:
+		//	  default1.bar.com
+		//	  foo0.bar.com: /foo0
+		//	  foo1.bar.com: /foo1
+		if numApps < numIng {
+			Failf("Need more apps than Ingress")
+		}
+		step := numApps / numIng
+		By(fmt.Sprintf("Creating %d rcs + svc, and %d routes per Ingress", numApps, step))
+		for i := 0; i < numApps; i = i + step {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: This logic is hard to follow.  Can you use more descriptive names than i and j to make it clearer?  Perhaps I'm being dumb, but it's not obvious to me what you're trying to do here.

File: test/e2e/ingress.go [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+		ings, err := client.Extensions().Ingress(ns).List(
+			labels.Everything(), fields.Everything())
+		Expect(err).NotTo(HaveOccurred())
+
+		for _, ing := range ings.Items {
+			// Wait for the loadbalancer IP.
+			start := time.Now()
+			address, err := waitForIngressAddress(client, ing.Namespace, ing.Name, lbPollTimeout)
+			Expect(err).NotTo(HaveOccurred())
+			By(fmt.Sprintf("Found address %v for ingress %v, took %v to come online",
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Admittedly other tests are not very consistent about this, but using a By() clause rather than a simple Log() here doesn't seem to make much sense?

File: test/e2e/ingress.go [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+				address, ing.Name, time.Since(start)))
+			creationTimes = append(creationTimes, time.Since(start))
+
+			// Check that all rules respond to a simple GET.
+			for _, rules := range ing.Spec.Rules {
+				if rules.IngressRuleValue.HTTP == nil {
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Under what conditions is this true?  Worth an explanatory comment?

File: test/e2e/ingress.go [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+					continue
+				}
+				for _, p := range rules.IngressRuleValue.HTTP.Paths {
+					route := fmt.Sprintf("http://%v%v", address, p.Path)
+					Logf("Testing route %v host %v with simple GET", route, rules.Host)
+
+					start := time.Now()
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: nit: I don't really like that you're re-using start both inside and outside this loop.

File: test/e2e/ingress.go [hunk 27]
```diff
@@ -0,0 +1,311 @@
...
+						var err error
+						lastBody, err = simpleGET(http.DefaultClient, route, rules.Host)
+						if err != nil {
+							Logf("host %v path %v: %v", rules.Host, route, err)
+							return false, nil
+						}
+						return true, nil
+					})
+					if pollErr != nil {
+						Failf("Last response body for %v, host %v:\n%v\n\n%v", route, rules.Host, lastBody, pollErr)
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Need to be more explicit here about what the error is, as it will look rather cryptic in the failure logs otherwise?  Perhaps something along the lines of "Expected successful GET within $lbPollTimout but didn't.  Last response for... "

File: test/e2e/resize_nodes.go [hunk 28]
```diff
@@ -111,25 +112,26 @@ func waitForGroupSize(size int) error {
 	return fmt.Errorf("timeout waiting %v for node instance group size to be %d", timeout, size)
 }
 
-func svcByName(name string) *api.Service {
+func svcByName(name string, port int) *api.Service {
```
Comment (quinton-hoole) [inline test/e2e/resize_nodes.go]: Can we move these two functions to e2e/util now that they're being used across multiple tests.  They're definitely not specific to resize_nodes.

File: test/e2e/serviceloadbalancers.go [hunk 29]
```diff
@@ -245,8 +245,13 @@ var _ = Describe("ServiceLoadBalancer", func() {
 })
 
 // simpleGET executes a get on the given url, returns error if non-200 returned.
-func simpleGET(c *http.Client, url string) (string, error) {
-	res, err := c.Get(url)
+func simpleGET(c *http.Client, url, host string) (string, error) {
```
Comment (quinton-hoole) [inline test/e2e/serviceloadbalancers.go]: Same comment as above.  Can we move this out of this test to a utility library.
Happy to defer to a separate PR if you prefer.  Just create a github issue and reference it here.

File: test/e2e/util.go [hunk 30]
```diff
@@ -2141,3 +2141,36 @@ func OpenWebSocketForURL(url *url.URL, config *client.Config, protocols []string
 	cfg.Protocol = protocols
 	return websocket.DialConfig(cfg)
 }
+
+// getIngressAddress returns the ips/hostnames associated with the Ingress.
+func getIngressAddress(client *client.Client, ns, name string) ([]string, error) {
```
Comment (quinton-hoole) [inline test/e2e/util.go]: I don't like that we're returning strings here.  Can we return something more strongly typed please?

File: test/e2e/util.go [hunk 31]
```diff
@@ -2141,3 +2141,36 @@ func OpenWebSocketForURL(url *url.URL, config *client.Config, protocols []string
...
+			addresses = append(addresses, a.IP)
+		}
+		if a.Hostname != "" {
+			addresses = append(addresses, a.Hostname)
+		}
+	}
+	return addresses, nil
+}
+
+// waitForIngressAddress waits for the Ingress to acquire an address.
```
Comment (quinton-hoole) [inline test/e2e/util.go]: Good to note what it's return values are.  And not have them be a string :-)

**Comment by quinton-hoole** [comment]
@bprashanth review done.

File: test/e2e/ingress.go (hunk 5 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: I had it at 5 previously but that's 180 lines of log spam. Thoughts?

File: test/e2e/ingress.go (hunk 13 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: This is just checking the times that are recorded in the actual tests. I want to log the percentiles even on failure. Ill move just the log lines to aftereach, and have the FailF in the actual test.

File: test/e2e/ingress.go (hunk 14 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: I think i'd have to encode it as Expect(numApps<numIng).NotToEqual(false) or something, which is weird

File: test/e2e/ingress.go (hunk 18 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: done, I didn't do that initially because usually the errors themselves have "Error: foo", so with enough call stacks you end up with "Error: Error: ..."

File: test/e2e/ingress.go (hunk 21 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: This isn't a failure actually, it's just polling on GLBC to cleanup resources. I think it'll be weird if it keeps printing "Failed to cleanup resources" 2/3 times and then passes. I'll add an assert.

File: test/e2e/ingress.go (hunk 23 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: Will clarify the variable names, the comment above was supposed to elucidate. Does it need clarification?

File: test/e2e/ingress.go (hunk 24 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: I use By when something of importance has occurred (in this case the loadbalancer acquiring an IP) because it prints out a bold STEP which sort of makes sense, it's the successful completion of a step in the test

File: test/e2e/ingress.go (hunk 25 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: never today, till we've exetended the api to include non-http Ingress

File: test/e2e/util.go (hunk 30 above)
Comment (bprashanth) [inline test/e2e/util.go]: It's actually just returning whatever the loadbalancer status has, which is a string because it could either be an IP or hostname and there's no type that captures both properly https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L1237

File: test/e2e/ingress.go (hunk 26 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: changed to GETStart

File: test/e2e/util.go (hunk 31 above)
Comment (bprashanth) [inline test/e2e/util.go]: Yeah unless you want me to invent some adhoc type for this i think we're stuck with the string in the api object because of the reason mentioned above

File: test/e2e/ingress.go (hunk 20 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: #16636

File: test/e2e/ingress.go (hunk 15 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: i now have expectedLBCreation/HealthCheck time.

File: test/e2e/ingress.go (hunk 17 above)
Comment (bprashanth) [inline test/e2e/ingress.go]: don't understand, do you mean I should clarify the comments I have above lbPollTimeout?

**Comment by bprashanth** [comment]
PTAL, I've addressed everything except:

Ill send a follow up pr for all the util/e2e functions you asked for
The last commit in my set has fixes for code review comments, will squash before merging
I'm also waiting on LGTM before pushing the image to gcr.io

File: hack/jenkins/e2e.sh [hunk 32]
```diff
@@ -164,6 +166,7 @@ GCE_SLOW_TESTS=(
 
 # Tests which are not able to be run in parallel.
 GCE_PARALLEL_SKIP_TESTS=(
+    "GCE\sL7\sLoadBalancer\sController"
```
Comment (quinton-hoole) [inline hack/jenkins/e2e.sh]: It's not clear to me that there is a real reason why these cannot run in parallel (other than just being a bit slow)?  Currently SLOW tests are explicitly skipped in PARALLEL runs (see line 314 in this file), so you don't need the above line, and can just omit it?

File: test/e2e/ingress.go (hunk 5 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Fair enough. Leave as 30 sec.

File: test/e2e/ingress.go (hunk 24 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Fair enough.

File: test/e2e/util.go (hunk 30 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: I'd prefer even a relatively dumb wrapper, along the lines you mention.  Anything other than a vanilla string would be an improvement.

File: test/e2e/util.go (hunk 31 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: My point was that the return value of this function needs a name, and/or a non-string type.

File: test/e2e/ingress.go (hunk 4 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Still TODO?

File: test/e2e/ingress.go (hunk 17 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: No, I was referring to numApps.  Why did you choose '6', as opposed to any other number? If someone came along and wanted to change the numbers, what would they need to know to do it intelligently?

File: test/e2e/ingress.go (hunk 26 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: nit: I realize that you're trying to echo the HTTP GET keyword's case here, but this is just plain wierd?  How about StartOfGet?

File: hack/jenkins/e2e.sh (hunk 32 above)
Comment (bprashanth) [inline hack/jenkins/e2e.sh]: They can't run in parallel with *themselves*, or other loadbalancer tests that use the same quota. I guess I was confused if this would ever happen. The second problem can be solved by simply requesting an increase (though oddly it looks like our slow-jenkins project is currently using 100% forwarding rules) as long as the first isn't a problem. Are you ok with bumping quotas way up (because we don't know how many L4 and L7 tests might run in parallel)?

File: test/e2e/ingress.go [hunk 33]
```diff
@@ -0,0 +1,315 @@
...
+	// Ingress as GCE is creating/checking backends in parallel, but at the
+	// same time, we're not testing GCE startup latency. So give it enough
+	// time, and fail if the average is too high.
+	lbPollTimeout  = 15 * time.Minute
+	lbPollInterval = 30 * time.Second
+
+	// One can scale this test by tweaking numApps and numIng, the former will
+	// create more RCs/Services and add them to a single Ingress, while the latter
+	// will create smaller, more fragmented Ingresses.
+	numApps = 6
```
Comment (quinton-hoole) [inline test/e2e/ingress.go]: To repeat a previous comment that might have been lost, why '6'?  The comment above does a reasonable job of describing what the knobs do, but not why you set them to 6 and 3 specifically.

File: test/e2e/util.go (hunk 31 above)
Comment (bprashanth) [inline test/e2e/util.go]: Not sure i understand, can you suggest something concrete? I'm pulling out strings from the apitype, i feel like this is not what you're asking for:

```
type addressStruct struct {
    addresses []string
}
```
but that's what it sounds like you're asking for :)

File: test/e2e/ingress.go (hunk 21 above)
Comment (quinton-hoole) [inline test/e2e/ingress.go]: Fair enough.

**Comment by quinton-hoole** [comment]
A few minor nits remaining, then lgtm.  Thanks for this @bprashanth !

File: hack/jenkins/e2e.sh (hunk 32 above)
Comment (quinton-hoole) [inline hack/jenkins/e2e.sh]: We don't currently run tests in parallel with themselves, anywhere, so don't worry about that (although I still don't quite understand why they could not logically run in parallel with themselves, provided that they run in their own namespace, which they do).

As for quota, yes, requesting additional quota for the relevant GCE projects is the right way to do (and, if you like, an explicit comment in this config, explaining that these tests are only disabled due to current lack of quota, rather than any real reason why they can't run in parallel) would be fine as a temporary solution until the quota increase goes though.

Comment (quinton-hoole) [inline hack/jenkins/e2e.sh]: Oh, and to be clear, PARALLEL_SKIP_TESTS is for tests that cannot run in parallel with *other* tests (e.g. tests that reboot nodes, resize clusters, saturate clusters with pods etc), because they break other tests.  Perhaps you want to update the comment accordingly.

Comment (bprashanth) [inline hack/jenkins/e2e.sh]: I just put in a quota request for more forwarding rules, though to make this parallel I'll have to bump it up even more, since we're currently at 100% forwarding rule usage (maybe an existing leak). The namespace doesn't matter in this case, as the L7 controller is running as a cluster addon (so it creates lbs for all namspaces) and the quota limitations are global. 

I'll add a todo

File: test/e2e/util.go (hunk 30 above)
Comment (bprashanth) [inline test/e2e/util.go]: Hmm Why? a list of strings is simple no? IMO a type is useful if it adds value, either in the form of validation or type checking. Cloaking a string in a type isn't much of an improvement?

**Comment by bprashanth** [comment]
PTAL. I've put in a request for increased quota in the k8s-jkns-e2e-gce-slow project, this test shouldn't run anywhere else because nothing has enough quota.

**Comment by bprashanth** [comment]
I've also put in a request for quota on kubernetes-jenkins, @quinton-hoole @kubernetes/goog-testing am I missing a project? this test cannot run in a project with the default quota settings

File: test/e2e/util.go (hunk 31 above)
Comment (quinton-hoole) [inline test/e2e/util.go]: Someone reading the function header would have no idea what the string return value would contain, unless they read through the code.

What I was suggesting is one of the following instead, approximately, in increasing order of my preference:

```
func waitForIngressAddress(c *client.Client, ns, ingName string, timeout time.Duration) (address string, error)
```

```
type address string
...
func waitForIngressAddress(c *client.Client, ns, ingName string, timeout time.Duration) (address, error)
```

**Comment by quinton-hoole** [comment]
Yes, according to the current config in jenkins/e2e.sh, the SLOW tests only run in the the *-slow GCE projects, e.g. 

https://github.com/kubernetes/kubernetes/blob/master/hack/jenkins/e2e.sh#L273

... unless the categorization of the test changes over time (e.g. gets tagged as FLAKY).

**Comment by bprashanth** [comment]
Hmm really, won't it run if you don't skip SLOW or focus !SLOW (i.e like in soak)?

**Comment by quinton-hoole** [comment]
Yes, you are absolutely right.  There are a few, like *-soak, and some of the release ones.  You'll need to increase the quota on all of these projects, or your test will fail in them.

**Comment by bprashanth** [comment]
Oh alright, I'll just go and submit 20 forms for quota now :)

**Comment by bprashanth** [comment]
I'm still waiting on the quota request to go through for our 20 odd projects

**Comment by bprashanth** [comment]
Alright quota request went through, shippable seems hosed. @quinton-hoole lgtm so when shippable recovers we can usher this in?

@ixdy @ihmccreery this is quota request I got yelled at for making ;)
Honestly there's no way around a blanket application of quota one test needs across several projects to guard against cases like this test getting demoted to flaky or running in parallel/in some release project etc, right? Maybe we can just use a single project with quota > 9000?

**Comment by quinton-hoole** [comment]
LGTM

**Comment by jlowdermilk** [comment]
@k8s-bot test this
@k8s-bot unit test this
